### PR TITLE
[PSL-108] B1 - Walletnode - Implement Sense API

### PR DESCRIPTION
**Pull request by bacnh8**
https://pastel.wiki/en/Architecture/OpenAPI/Sense#walletnode

Step1-03/04: Implement Sense API /openapi/sense/upload
Step1-05/11: Implement Sense API /openapi/sense/details/{task_id}

### CODE REVIEW

File: pastel/client.go [hunk 1]
```diff
@@ -382,6 +382,27 @@ func (client *client) GetRegisterNFTFee(ctx context.Context, request GetRegister
...
+	}
+
+	var actionFee struct {
+		ActionFee int64 `json:"actionfee"`
+	}
+
+	params := []interface{}{}
+	params = append(params, "getactionfee")
+	params = append(params, actionType)
+	params = append(params, ImgSizeInMb)
```
Comment (a-ok123) [inline pastel/client.go:397]: API was modified. call:
`getactionfees <data size>`

response:
```
{
  "sensefee": x,
  "cascadefee": y
}
```

**Review by a-ok123 — CHANGES_REQUESTED** [review]

File: walletnode/api/design/openapi.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	//revive:disable:dot-imports
+	//lint:ignore ST1001 disable warning dot import
+	. "goa.design/goa/v3/dsl"
+	//revive:enable:dot-imports
+
+	cors "goa.design/plugins/v3/cors/dsl"
+)
+
+var _ = Service("openapi", func() {
+	Description("OpenAPI Sensei service")
```
[the remark this answers is not in the document]
Comment (bacnh8) [reply walletnode/api/design/openapi.go]: This is already improved in the latest commit.

File: walletnode/api/services/openapi.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+func NewOpenAPI() *OpenAPI {
+	return &OpenAPI{
+		Common:   NewCommon(),
+		imageTTL: defaultImageTTL,
+	}
+}
+
+// OpenApiUploadImageDecoderFunc implements the multipart decoder for service "artworks" endpoint "UploadImage".
+// The decoder must populate the argument p after encoding.
+func OpenApiUploadImageDecoderFunc(ctx context.Context, service *OpenAPI) server.OpenapiUploadImageDecoderFunc {
```
[the remark this answers is not in the document]
Comment (bacnh8) [reply walletnode/api/services/openapi.go]: This is already improved in the latest commit.

File: pastel/client.go (hunk 1 above)
Comment (bacnh8) [reply pastel/client.go:397]: ok, I will improve pastel API call with the latest commit.

File: walletnode/services/senseregister/ticket.go [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package senseregister
+
+import "github.com/pastelnetwork/gonode/common/service/artwork"
+
+type Request struct {
```
Comment (nguyen202) [inline walletnode/services/senseregister/ticket.go:5]: nit: have to add description for this structure

File: walletnode/api/gen/http/artworks/client/client.go [hunk 5]
```diff
@@ -142,9 +142,7 @@ func (c *Client) RegisterTaskState() goa.Endpoint {
 			return nil, err
 		}
 		var cancel context.CancelFunc
-		_, cancel = context.WithCancel(ctx)
-		defer cancel()
-
+		ctx, cancel = context.WithCancel(ctx)
```
Comment (nguyen202) [inline walletnode/api/gen/http/artworks/client/client.go:145]: this is bug of GOA - with this generated code - it call deferred function (`cancel()`) immediately and stop request immediately - while we want to listen the change in task state. please revert this back to :
```
_, cancel = context.WithCancel(ctx)
defer cancel()
```

File: walletnode/api/gen/http/artworks/client/client.go [hunk 6]
```diff
@@ -248,9 +246,7 @@ func (c *Client) ArtSearch() goa.Endpoint {
...
```
Comment (nguyen202) [inline walletnode/api/gen/http/artworks/client/client.go:249]: same above:
this is bug of GOA - with this generated code - it call deferred function (cancel()) immediately and stop request immediately - while we want to listen the change in task state. please revert this back to :
```
_, cancel = context.WithCancel(ctx)
defer cancel()
````

File: walletnode/api/services/multipart.go [hunk 7]
```diff
@@ -30,57 +31,68 @@ func UploadImageDecoderFunc(ctx context.Context, service *Artwork) server.Artwor
...
-			part, err := reader.NextPart()
-			if err != nil {
-				if err == io.EOF {
-					break
-				}
-				return artworks.MakeInternalServerError(errors.Errorf("could not read next part: %w", err))
-			}
+		filename, err := handleUploadImage(ctx, reader, service.register.Storage)
+		if err != nil {
+			return artworks.MakeBadRequest(err)
```
Comment (nguyen202) [inline walletnode/api/services/multipart.go:36]: prefer old implementation with more details of type of error - new your changes always return BadRequest error - while some other problems (like storage full is internal server error).

File: walletnode/api/services/sense.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+	return res, nil
+}
+
+// StartTask - Starts a action data task
+func (service *Sense) ActionDetails(ctx context.Context, p *sense.ActionDetailsPayload) (res *sense.ActionDetailResult, err error) {
+	// get image filename from storage based on image_id
+	filename, err := service.db.Get(p.ImageID)
+	if err != nil {
+		return nil, sense.MakeInternalServerError(err)
```
Comment (nguyen202) [inline walletnode/api/services/sense.go:80]: could you wrap error with context?
like : sense.MakeInternalServerError(errors.Errorf("get image filename: %w", err))

File: walletnode/api/services/sense.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	// get image data from storage
+	file, err := service.register.Storage.File(string(filename))
+	if err != nil {
+		return nil, sense.MakeInternalServerError(err)
```
Comment (nguyen202) [inline walletnode/api/services/sense.go:86]: could you wrap error with context?
like : sense.MakeInternalServerError(errors.Errorf("get image file: %w", err))

File: walletnode/services/senseregister/service.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+func (service *Service) AddTask(ticket *Request) string {
+	task := NewTask(service, ticket)
+	service.Worker.AddTask(task)
+
+	return task.ID()
+}
+
+// VerifyImageSignature verifies the signature of the image
+func (service *Service) VerifyImageSignature(ctx context.Context, file *artwork.File, signature string, pastelID string) error {
+	imgData, err := file.Bytes()
```
Comment (nguyen202) [inline walletnode/services/senseregister/service.go:42]: should verify if file is nil

Comment (bacnh8) [reply walletnode/services/senseregister/service.go:42]: File is openned before that, I will improve to re-used imgdata instead.

**Review by nguyen202 — APPROVED** [review]

**Pull request merged**